File: tmp/sys/SwordOffer/src/main/scala/com/scala/offer/Problem7/MyStack.scala
package com.scala.offer.Problem7

import java.util

/**
 * Created by Lemonjing on 2018/5/8.
 * Github: Lemonjing
 *
 * 两个队列实现栈
 */
object MyStack {
  private val queue1 = new util.LinkedList[Integer]()
  private val queue2 = new util.LinkedList[Integer]()

  def push(e: Integer): Unit = {
    if (!queue1.isEmpty) {
      queue1.offer(e)
    } else if (!queue2.isEmpty) {
      queue2.offer(e)
    } else {
      queue1.offer(e)
    }
  }

  def pop(): Integer = {
    if (queue1.isEmpty && queue2.isEmpty) {
      throw new Exception("error, stack is empty.")
    }
    if (!queue1.isEmpty) {
      var n = queue1.size()
      while (n > 1) {
        queue2.offer(queue1.poll())
        n -= 1
      }
      return queue1.poll()
    } else {
      var n = queue2.size()
      while (n > 1) {
        queue1.offer(queue2.poll())
        n -= 1
      }
      return queue2.poll()
    }
  }
}

File: tmp/sys/SwordOffer/src/main/java/com/rann/offer/problem7/MyStack.java
package com.rann.offer.problem7;

import java.util.LinkedList;
import java.util.Queue;

/**
 * 两个队列实现一个栈
 *
 * @author lemonjing
 * @version 1.0
 */
public class MyStack {
    private Queue<Integer> q1 = new LinkedList<>();
    private Queue<Integer> q2 = new LinkedList<>();

    public Integer push(Integer a) {
        if (q1.isEmpty() && !q2.isEmpty()) {
            q2.offer(a);
        }
        if (!q1.isEmpty() && q2.isEmpty()) {
            q1.offer(a);
        }
        //都空时从q1开始
        if (q1.isEmpty() && q2.isEmpty()) {
            q1.offer(a);
        }
        return a;
    }

    public Integer pop() throws Exception {
        Integer res = null;
        if (q1.isEmpty() && q2.isEmpty()) {
            throw new Exception("Error,栈为空");
        }

        if (q1.isEmpty() && !q2.isEmpty()) {
            int i = q2.size();
            while (i > 1) {
                q1.offer(q2.poll());
                i--;
            }
            res = q2.poll();
        } else if (!q1.isEmpty() && q2.isEmpty()) {
            int i = q1.size();
            while (i > 1) {
                q2.offer(q1.poll());
                i--;
            }
            res = q1.poll();
        }

        return res;
    }
}
